policy: ensure Endpoint lock held while accessing identity

ianvernon opened the pull request:
Because the endpoint identity is mutable, we need to hold the lock for the endpoint while we are updating the caches in the rules which specify what identities are selected by that rule. The endpoint's identity may change while we are updating the caches, which could result in an identity being left behind in said caches. Also, the endpoint's identity may be set to a nil value, which would result in a panic, since fields within the identity are dereferenced to access the identity's labels when the caches within the rules are updated.

The locks which check liveness with the endpoint are used here because if the endpoint is being deleted, we don't need to update the caches within the rules anymore for it. Also, when an endpoint is deleted, its security identity is set to `nil` in `pkg/endpoint.go:LeaveLocked`. Read-locking is needed because no updating of the security identity is performed here.

There is still a possibility that the endpoint's identity could be nil while itis 'alive', as the endpoint is inserted into the endpointmanager before its identity is allocated in `daemon/endpoint.go:createEndpoint`. This is OK, as once an endpoint's identity is changed, another regeneration will occur for it, which will populate the caches within the rules with its new identity. This case, while possible, is not probable.
    
Signed-off by: Ian Vernon <ian@cilium.io>

Related: #7837

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7844)
<!-- Reviewable:end -->

Commit:
endpoint: fix comment for GetSecurityIdentity

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add RLockAlive, RUnlock to Endpoint interface

This is done to ensure that the Endpoint's security identity is not concurrently
accessed by threads which update the caches in the rules and threads which
change the identity of the endpoint itself.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: ensure Endpoint lock held while accessing identity

Because the endpoint identity is mutable, we need to hold the lock for the
endpoint while we are updating the caches in the rules which specify what
identities are selected by that rule. The endpoint's identity may change while
we are updating the caches, which could result in an identity being left behind
in said caches. Also, the endpoint's identity may be set to a nil value, which
would result in a panic, since fields within the identity are dereferenced to
access the identity's labels when the caches within the rules are updated.

The locks which check liveness with the endpoint are used here because if the
endpoint is being deleted, we don't need to update the caches within the rules
anymore for it. Also, when an endpoint is deleted, its security identity is set
to `nil` in `pkg/endpoint.go:LeaveLocked`. Read-locking is needed because no
updating of the security identity is performed here.

There is still a possibility that the endpoint's identity could be nil while it
is 'alive', as the endpoint is inserted into the endpointmanager before its
identity is allocated in `daemon/endpoint.go:createEndpoint`. This is OK, as
once an endpoint's identity is changed, another regeneration will occur for it,
which will populate the caches within the rules with its new identity. This
case, while possible, is not probable.

Signed-off by: Ian Vernon <ian@cilium.io>

nebril commented on pkg/policy/rules.go:
```diff
@@ -359,13 +360,21 @@ loop:
+	if err := ep.RLockAlive(); err != nil {
+		return false, fmt.Errorf("cannnot update caches in rules for endpoint %d because it is being deleted: %s", id, err)
+	}
+	defer ep.RUnlock()
```
Do we want to release the lock right after obtaining the security identity? If yes, we might want to put `securityIdentity` retrieval into inline closure func with deferred unlock.

joestringer replied:
There's another case here which is that when an endpoint is deleted causing an identity to no longer be used locally, the identitymanager will call out into the policy package to flush that identity out from all of the rule caches here. We want to make sure that this `updateEndpointsCaches()` case finishes all of its work before that other side is triggered, otherwise there's a race where:

1. The policy is updated, causing a call to `updateEndpointsCaches()`
1. We get the Identity here
1. The endpoint is removed
1. The endpoint removal causes the policy repository to get updated to remove this Identity from these caches (note: This does not take the policy repo write lock right now because it's only changing the rule caches)
1. Finally, this code path adds the Identity into the rule caches

This would result in Identities in the rule caches to exist even though the identity isn't locally used by any endpoint. We would never clean this up.

(Now, maybe there's another solution to this particular problem but at least for now it's probably safer to keep this Endpoint RLock to block out endpoint deletion until the caches are updated from the policy update side here).

joestringer replied:
Actually, might be worth double-checking (4) above. Surely we *should* be taking a readlock to go over the repository, which would automatically solve this already..

EDIT: And yep, actually we take the readlock so the above case is totally not a problem:

https://github.com/cilium/cilium/blob/9765d3adffd1554793aeee123a80b23e17b5b6ed/pkg/policy/repository.go#L430-L433

ianvernon replied:
@joestringer I'm not 100% sure that taking the `RLock` of the repository there guarantees the race might not occur, because the policy repository lock is not always held when `updateEndpointsCaches` due to goroutines being spawned to perform the work of updating the caches.

See: 

https://github.com/cilium/cilium/blob/9765d3adffd1554793aeee123a80b23e17b5b6ed/daemon/policy.go#L378-L380

 We have a WaitGroup passed into the funtion from `policyAdd` and `policyDelete` which we could wait on *before* releasing the repository mutex?  I did this intentionally when I wrote the selective regeneration work since I didn't want to hold the lock of the repository itself, since the rule content itself wasn't being changed, rather the caches themselves, which have their own mutex. I think holding the endpoint mutex for the duration of this is OK, as keeping one endpoint locked longer vs. the policy repository is better, since the repository is a global structure. 

All of these mutexes are making my head hurt 😢

joestringer replied:
@ianvernon Hmm, interesting point.... is that safe? Shouldn't we at least hold the repository RLock while iterating rules in the repository during cache update, to ensure that two policy changes in quick succession don't end up causing havoc with the the cache update goroutines?

ianvernon replied:
@joestringer what sort of havoc were you envisioning? The policy addition logic itself is protected by the queueing logic (policy repository event queues), so we should be insulated from such a case. A policy computation for an endpoint could occur while we are updating the caches, though, but that's OK because the content of the rules themselves is not changing.

My goal here was to reduce contention on the policy repository lock since it's used across the entire system. I do think that this logic is far too complicated, though. Too many means of serialization makes it impossible to reason about how things work =/

joestringer replied:
My question is with the following order of events:

1. Policy update A comes in, is queued into `RepositoryChangeQueue`
1. Policy update B comes in, is queued into `RepositoryChangeQueue`
1. Take A off the `RepositoryChangeQueue`, grab the mutex
1. Make policy changes for A
1. Trigger selection cache updates for A
1. Release the mutex
1. Add events to handle regerenation for A to `RuleReactionQueue`
1. Take B off the `RepositoryChangeQueue`, grab the mutex
1. Start processing changes for B
1. ...

How do you know that (8) blocks on the selection update completing in goroutines triggered from (5)?

I see a policySelectionWG, but that seems at a glance to be passed down to the rule reaction queue to ensure that the processing associated with (7) doesn't happen until after (5)'s work is complete. However AFAICT it doesn't say anything about ensuring that (5)'s work completes before (9) starts manipulating the repository.

ianvernon replied:
Discussed offline. The operation `UpdateRulesEndpointsCaches`, which updates the caches, is performed upon slices of pointers to rules from the policy repository. Not the repository itself. So, we do not have to worry about the policy repository being manipulated while these slices are being analyzed.

nebril reviewed (commented):
LGTM overall, but please check the comment inline.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -456,7 +456,16 @@ func (r ruleSlice) UpdateRulesEndpointsCaches(endpointsToBumpRevision *EndpointS
+		// endpointsToRegenerate, as we could not determine what to do for the
+		// endpoint). This is usually the case when the endpoint is no longer
+		// alive (i.e., it has been marked to be deleted).
+		if endpointSelected || err != nil {
```
This new error is not being logged anywhere, did you mean to at least put this in as a debug log?

joestringer reviewed (approved):
Minor nit.

One other query, this will block out some endpoint interactions whenever policy is being changed for the duration of an iteration over the policy repository (albeit, not policy computation so not super expensive). Do we have appropriate logging of delays in the other paths? For instance, when deleting an endpoint?

I suppose that the metrics around locking will provide us visibility into this for "free" in the logs.

Commit:
policy: add debug log when error from `updateEndpointsCaches` is non-nil

This error is harmless for the most part, but means that an endpoint has been
deleted or has a nil identity while we are still trying to update the caches in
the rules for it, which is not optimal functionality.

Signed-off by: Ian Vernon <ian@cilium.io>
